File: capstoneproject-to-do-fe/To-Do-App/src/app/services/todo.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { USER } from '../model/user';
import { ToDo } from '../model/todo';
import { LoginService } from './login.service';

@Injectable({
  providedIn: 'root',
})
export class TodoService {
  //inject HttpClient service in constructor to call backend services
  constructor(private http: HttpClient, private loginService: LoginService) {}

  todoList: ToDo[] = [];

  //created addToDo()
  public addToDo(toDo: any): Observable<any> {
    return this.http.post<any>(`http://localhost:9005/toDo/user/addToDo`, toDo);
  }

  //created getAllToDo()
  public getAllToDo(): Observable<any> {
    // const jwtToken = this.loginService.getToken();
    // let header = new HttpHeaders();
    // if (jwtToken !== null) {
    //   header = new HttpHeaders({
    //     Authorization: `Bearer${jwtToken}`,
    //   });
    // }
    // return this.http.get<any>(`http://localhost:9005/toDo/user/getAllToDo`, {
    //   headers: header,
    // });
    return this.http.get<any>(`http://localhost:9005/toDo/user/getAllToDo`);
  }

  //created updateToDo()
  public updateToDo(toDo: any): Observable<any> {
    return this.http.put<any>(
      `http://localhost:9005/toDo/user/updateToDo`,
      toDo
    );
  }

  //created deleteToDoByTitle()
  public deleteToDoByTitle(title: any): Observable<any> {
    return this.http.delete<any>(
      `http://localhost:9005/toDo/user/deleteToDoByTitle/${title}`
    );
  }

  //created getToDoByTitle()
  public getToDoByTitle(title: any): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/toDo/user/getToDoByTitle/${title}`
    );
  }

  //create getAllListByPriority()
  public getAllListByPriority(priority: any): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/toDo/user/getAllListByPriority/${priority}`
    );
  }

  //create getAllToDoByArchiveStatus()
  public getAllToDoByArchiveStatus(isArchive: any): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/toDo/user/getAllToDoByArchiveStatus/${isArchive}`
    );
  }

  //create updateTaskArchiveStatus()
  public updateTaskArchiveStatus(toDo: ToDo): Observable<Array<ToDo>> {
    return this.http.put<Array<ToDo>>(
      `http://localhost:9005/toDo/user/updateTaskArchiveStatus`,
      toDo
    );
  }

  //create updateToDoStatus()
  public updateToDoStatus(toDo: any): Observable<any> {
    return this.http.put<any>(
      `http://localhost:9005/toDo/user/updateToDoStatus`,
      toDo
    );
  }

  //create getAllToDoByStatus()
  public getAllToDoByStatus(isCompleted: any): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/toDo/user/getAllToDoByStatus/${isCompleted}`
    );
  }

  //create getAllToDoByStatus()
  public getToDoByImportantStatus(isCompleted: any): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/toDo/user/getToDoByImportantStatus/${isCompleted}`
    );
  }

  //created deleteTaskByTitle()
  public deleteTaskByTitle(title: any): Observable<any> {
    return this.http.delete<any>(
      `http://localhost:9005/toDo/user/deleteToDoByTitle/${title}`
    );
  }

  //created getAllNotification()
  public getAllNotification(): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/notificationController/user/getAllNotification`
    );
  }

  public updateNotification(toDo: any): Observable<any> {
    return this.http.put<any>(
      `http://localhost:9005/notificationController/user/updateNotification`,
      toDo
    );
  }

  //created deleteAllToDo()
  public deleteAllToDo(): Observable<any> {
    return this.http.delete<any>(
      `http://localhost:9005/toDo/user/deleteAllToDo`
    );
  }

  public getToDOByCompletedStatus(isCompleted: boolean): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/toDo/user/getAllToDoByStatus/${isCompleted}`
    );
  }

  //create updateTaskArchiveStatus()
  public updateTaskAsCompletedTask(toDo: any): Observable<any> {
    return this.http.put<any>(
      `http://localhost:9005/toDo/user/updateToDoAsCompletedTask`,
      toDo
    );
  }

  //create updateToDoAsImportantTask()
  public updateToDoAsImportantTask(toDo: any): Observable<any> {
    return this.http.put<any>(
      `http://localhost:9005/toDo/user/updateToDoAsImportantTask`,
      toDo
    );
  }

  //getCurrentUser
  public getCurrentUser(): Observable<USER> {
    return this.http.get<USER>(
      `http://localhost:9005/toDo/user/getUserDetails`
    );
  }

  public getToDOByDate(): Observable<any> {
    return this.http.get<any>(`http://localhost:9005/toDo/user/getToDoByDate`);
  }

  //create getAllToDoByStatus()
  public getToDoByPersonalStatus(isCompleted: any): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/toDo/user/getToDoByPersonalStatus/${isCompleted}`
    );
  }

  public updateToDoAsPersonalTask(toDo: any): Observable<any> {
    return this.http.put<any>(
      `http://localhost:9005/toDo/user/updateToDoAsPersonalTask`,
      toDo
    );
  }
  //create getAllToDoByStatus()
  public getToDoByTravelStatus(isCompleted: any): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/toDo/user/getToDoByTravelStatus/${isCompleted}`
    );
  }

  public updateToDoAsTravelTask(toDo: any): Observable<any> {
    return this.http.put<any>(
      `http://localhost:9005/toDo/user/updateToDoAsTravelTask`,
      toDo
    );
  }

  public addCustomCategory(Category: any): Observable<any> {
    return this.http.post<any>(
      'http://localhost:9005/toDo/user/createCategory',
      Category
    );
  }

  public getAllCategory(): Observable<any> {
    return this.http.get<any>('http://localhost:9005/toDo/user/getAllCategory');
  }

  public getAllCatergoryByToDoList(title: string): Observable<any> {
    return this.http.get<any>(
      `http://localhost:9005/toDo/user/getAllCatergoryByToDoList/${title}`
    );
  }

  public addTodoToUserCategory(title: string, toDo: any): Observable<any> {
    return this.http.post<any>(
      `http://localhost:9005/toDo/user/addTodoToUserCategory/${title}`,
      toDo
    );
  }
}

File: to_do_Backend/ToDoService/src/main/java/com/niit/bej/todoservice/service/ToDoService.java
package com.niit.bej.todoservice.service;

import com.niit.bej.todoservice.domain.Category;
import com.niit.bej.todoservice.domain.ToDo;
import com.niit.bej.todoservice.domain.User;
import com.niit.bej.todoservice.exception.*;

import java.util.List;

public interface ToDoService {
    User registerUser(User user) throws UserAlreadyExistException;

    User addToDo(String emailId, ToDo toDo) throws UserNotFoundException, ToDoAlreadyExistException;

    List<ToDo> getAllToDo(String emailId) throws UserNotFoundException, ToDoNotFoundException;

    User updateToDo(String emailId, ToDo toDo) throws UserNotFoundException, ToDoNotFoundException;

    boolean deleteToDoByTitle(String emailId, String title) throws UserNotFoundException, ToDoNotFoundException;

    ToDo getToDoByTitle(String emailId, String title) throws UserNotFoundException, ToDoNotFoundException;

    List<ToDo> getAllToDoByPriority(String emailId, String priority) throws UserNotFoundException;

    List<ToDo> getToDoListByArchivedStatus(String emailId, boolean isArchive) throws UserNotFoundException, ToDoNotFoundException;

    User updateTaskAsArchiveStatus(String emailId, ToDo toDo) throws UserNotFoundException, ToDoNotFoundException;

    List<ToDo> getToDoByCompletionStatus(String emailId, Boolean isCompleted) throws ToDoNotFoundException, UserNotFoundException;

    User updateToDoAsCompletedTask(String emailId, ToDo toDo) throws ToDoNotFoundException, UserNotFoundException;

    List<ToDo> getToDoByImportantStatus(String emailId, Boolean isCompleted) throws ToDoNotFoundException, UserNotFoundException;

    User updateToDoAsImportantTask(String emailId, ToDo toDo) throws ToDoNotFoundException, UserNotFoundException;

    boolean deleteAllToDo(String emailId) throws UserNotFoundException, ToDoNotFoundException;

    User getUserDetails(String emailId) throws UserNotFoundException;

    List<ToDo> getAllToDoByDate(String emailId) throws UserNotFoundException, ToDoNotFoundException;

    User updateToDoAsTravelTask(String emailId, ToDo toDo) throws ToDoNotFoundException, UserNotFoundException;

    List<ToDo> getToDoByTravelStatus(String emailId, Boolean isCompleted) throws ToDoNotFoundException, UserNotFoundException;

    List<ToDo> getToDoByPersonalStatus(String emailId, Boolean isCompleted) throws ToDoNotFoundException, UserNotFoundException;

    User updateToDoAsPersonalTask(String emailId, ToDo toDo) throws ToDoNotFoundException, UserNotFoundException;

    User CreateCategorylist (Category category, String emailId) throws UserNotFoundException, CategoryAlreadyExistException;
    List<Category> getAllCatergory(String emailId) throws UserNotFoundException;

    Category saveUserTodoToCategory(String title, String emailId, ToDo toDo) throws UserNotFoundException, ToDoNotFoundException, ToDoAlreadyExistException;
    public List<Category> getAllCatergoryByToDoList(String emailId, String title)throws UserNotFoundException, CategoryNotFoundException;


}
